native/src/main/java/io/ballerina/stdlib/log/MaskedStringBuilder.java

```diff
@@ -0,0 +1,574 @@
+            RecordType recType = (RecordType) valueType;
+            fields = recType.getFields();
+            // Use cached field annotations for better performance
+            fieldAnnotations = getCachedFieldAnnotations(recType);
+        }
+
+        return processRecordValue(mapValue, fieldAnnotations, fields);
+    }
+
+    private String processRecordValue(BMap<?, ?> mapValue, Map<String, BMap<?, ?>> fieldAnnotations,
+                                      Map<String, Field> fields) {
+        int startPos = stringBuilder.length();
+
+        stringBuilder.append('{');
+        addRecordFields(mapValue, fields, fieldAnnotations);
+        stringBuilder.append('}');
+
+        String result = stringBuilder.substring(startPos);
+        stringBuilder.setLength(startPos);
+        return result;
+    }
+
+    private void addRecordFields(BMap<?, ?> mapValue, Map<String, Field> fields,
+                                 Map<String, BMap<?, ?>> fieldAnnotations) {
+        boolean first = true;
+
+        for (Object key : mapValue.getKeys()) {
+            if (!(key instanceof BString keyStr)) {
+                continue;
+            }
+            Object fieldValue = mapValue.get(key);
+            String fieldName = keyStr.getValue();
+            first = fields.containsKey(fieldName) ?
+                    addDefinedFieldValue(fieldAnnotations, fieldName, fieldValue, first) :
+                    addDynamicFieldValue(fieldValue, first, fieldName);
+        }
+    }
+
+    private boolean addDynamicFieldValue(Object fieldValue, boolean first, String fieldName) {
+        String fieldStringValue = buildInternal(fieldValue);
+        if (!first) {
+            stringBuilder.append(',');
+        }
+        appendFieldToJsonOptimized(fieldName, fieldStringValue, false, fieldValue);
+        return false;
+    }
+
+    private boolean addDefinedFieldValue(Map<String, BMap<?, ?>> fieldAnnotations, String fieldName, Object fieldValue,
+                                         boolean first) {
+        Optional<BMap<?, ?>> annotation = getLogSensitiveDataAnnotation(fieldAnnotations, fieldName);
+        Optional<String> fieldStringValue = annotation
+                .map(fieldAnnotation -> getStringValue(fieldAnnotation, fieldValue, runtime))
+                .orElseGet(() -> Optional.of(buildInternal(fieldValue)));
+
+        if (fieldStringValue.isPresent()) {
+            if (!first) {
+                stringBuilder.append(',');
+            }
+            appendFieldToJsonOptimized(fieldName, fieldStringValue.get(), annotation.isPresent(), fieldValue);
+            first = false;
+        }
+        return first;
+    }
+
+    /**
+     * Optimized version of appendFieldToJson that writes directly to StringBuilder
+     * without creating intermediate String objects for better performance.
+     */
+    private void appendFieldToJsonOptimized(String fieldName, String value, boolean hasAnnotation, Object fieldValue) {
```

Comment:
Do we need the `optimized` part in the method names? Check similar places.

Reply:
Fixed via [37603fb](https://github.com/ballerina-platform/module-ballerina-log/pull/1343/commits/37603fbb99b5a2cec03f41ccd367ec65e3152956)